clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java

```diff
@@ -90,6 +91,15 @@ public int count() {
         return count;
     }
 
+    public Map<TopicPartition, OffsetAndMetadata> nextOffsets() {
```

Comment:
This doesn't seem to achieve the goal that we are intending. Because the offset that we want to commit is exactly not `lastRecordOffset + 1`, but the position inside the consumer. I think we need to pass the OffsetAndMetadata into this class as part of the constructor.

That would mean we need to, after all, propose a KIP that extends the constructor of this class, because the constructor is public. Sorry, I missed that before. Let's implement it first, before we write the KIP, just to make sure that we get it right this time!

Reply:
Good callout!

Also wondering if `lastRecord.leaderEpoch()`is what we want? Or should it be `nextRecord.leaderEpoch()` ?

Reply:
The position inside the consumer is `nextFetchPosition` with leader epoch `lastEpoch`, so I think we want to commit that.